hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java

```diff
@@ -129,6 +133,123 @@ public static OMPathInfo verifyFilesInPath(
     return new OMPathInfo(missing, OMDirectoryResult.NONE, inheritAcls);
   }
 
+  /**
+   * Verify any dir/key exist in the given path in the specified
+   * volume/bucket by iterating through directory table.
+   *
+   * @param omMetadataManager OM Metadata manager
+   * @param volumeName        volume name
+   * @param bucketName        bucket name
+   * @param keyName           key name
+   * @param keyPath           path
+   * @return OMPathInfoV1 path info object
+   * @throws IOException on DB failure
+   */
+  public static OMPathInfoV1 verifyDirectoryKeysInPath(
```

Comment:
Modified logic below, looks like few cases are missed.

Have a look into it, and reuse it as needed.


```
  public static OMPathInfoV1 verifyDirectoryKeysInPath(
          @Nonnull OMMetadataManager omMetadataManager,
          @Nonnull String volumeName,
          @Nonnull String bucketName, @Nonnull String keyName,
          @Nonnull Path keyPath) throws IOException {

    String leafNodeName = OzoneFSUtils.getFileName(keyName);
    List<String> missing = new ArrayList<>();
    List<OzoneAcl> inheritAcls = new ArrayList<>();
    OMDirectoryResult result = OMDirectoryResult.NONE;

    Iterator<Path> elements = keyPath.iterator();
    // TODO: volume id and bucket id generation logic.
    String bucketKey = omMetadataManager.getBucketKey(volumeName, bucketName);
    long lastKnownParentId = omMetadataManager.getBucketTable().get(bucketKey)
        .getObjectID();
    String dbDirName = ""; // absolute path for trace logs
    while (elements.hasNext()) {
      String fileName = elements.next().toString();
      if (missing.size() > 0) {
        // Add all the sub-dirs to the missing list except the leaf element.
        // For example, /vol1/buck1/a/b/c/d/e/f/file1.txt.
        // Assume /vol1/buck1/a/b/c exists, then add d, e, f into missing list.
        if(elements.hasNext()){
          // skips leaf node.
          missing.add(fileName);
        }
        continue;
      }

      // For example, /vol1/buck1/a/b/c/d/e/f/file1.txt
      // 1. Do lookup on directoryTable. If not exists goto next step.
      // 2. Do look on keyTable. If not exists goto next step.
      // 3. Add 'sub-dir' to missing parents list
      String dbNodeName = omMetadataManager.getOzoneLeafNodeKey(
              lastKnownParentId, fileName);
      OmDirectoryInfo omPrefixInfo = omMetadataManager.getDirectoryTable().
              get(dbNodeName);
      if (omPrefixInfo != null) {
        dbDirName += omPrefixInfo.getName() + OzoneConsts.OZONE_URI_DELIMITER;
        if (elements.hasNext()) {
          result = OMDirectoryResult.DIRECTORY_EXISTS_IN_GIVENPATH;
          lastKnownParentId = omPrefixInfo.getObjectID();
          inheritAcls = omPrefixInfo.getAcls();
          continue;
        } else {
          // Checked all the sub-dirs till the leaf node.
          // Found a directory in the given path.
          result = OMDirectoryResult.DIRECTORY_EXISTS;
          break;
        }
      } else {
        if (omMetadataManager.getKeyTable().isExist(dbNodeName)) {
          if (elements.hasNext()) {
            // Found a file in the given key name.
            result = OMDirectoryResult.FILE_EXISTS_IN_GIVENPATH;
          } else {
            // Checked all the sub-dirs till the leaf file.
            // Found a file with the given key name.
            result = OMDirectoryResult.FILE_EXISTS;
          }
          break; // Skip directory traversal as it hits key.
        }
      }

      // Add to missing list, there is no such key/directory with given name
      if (elements.hasNext()) {
        missing.add(fileName);
      }
    }

    if (result == OMDirectoryResult.DIRECTORY_EXISTS_IN_GIVENPATH) {
      String dbDirKeyName = omMetadataManager.getOzoneDirKey(volumeName,
          bucketName, dbDirName);
      LOG.trace("Acls inherited from parent " + dbDirKeyName + " are : "
          + inheritAcls);
    }

    if (result != OMDirectoryResult.NONE) {
      LOG.trace("verifyFiles in Path : " + "/" + volumeName
              + "/" + bucketName + "/" + keyName + ":" + result);
      return new OMPathInfoV1(leafNodeName, lastKnownParentId, missing, result,
              inheritAcls, dbDirName);
    }

    if (inheritAcls.isEmpty()) {
      inheritAcls = omMetadataManager.getBucketTable().get(bucketKey)
              .getAcls();
      LOG.trace("Acls inherited from bucket " + bucketName + " are : "
              + inheritAcls);
    }

    LOG.trace("verifyFiles in Path : " + volumeName + "/" + bucketName + "/"
            + keyName + ":" + result);

    // Found no files/ directories in the given path.
    return new OMPathInfoV1(leafNodeName, lastKnownParentId, missing,
            OMDirectoryResult.NONE, inheritAcls, null);
  }
```

Reply:
Thanks for the help. Will incorporate it.